File: plan/parser.py
"""Markdown parser for project plans."""

from __future__ import annotations

import re
from datetime import datetime
from pathlib import Path
from typing import Optional

from .models import (
    AcceptanceCriterion,
    TestScenario,
    Task,
    TaskStatus,
    Story,
    StoryStatus,
    Epic,
    EpicStatus,
    Milestone,
    Persona,
    ProjectPlan,
)


class PlanParser:
    """Parses markdown files into plan data structures."""

    def parse_plan(self, plan_dir: Path) -> ProjectPlan:
        """Parse a complete project plan from a directory."""
        plan_dir = Path(plan_dir)

        # Parse project.md for overview
        project_file = plan_dir / "project.md"
        name, description, milestones = self._parse_project_file(project_file)

        # Parse personas
        personas_dir = plan_dir / "personas"
        personas = self._parse_personas_dir(personas_dir)

        # Parse epics and stories
        epics_dir = plan_dir / "epics"
        stories_dir = plan_dir / "stories"
        epics = self._parse_epics_and_stories(epics_dir, stories_dir)

        return ProjectPlan(
            name=name,
            description=description,
            epics=epics,
            milestones=milestones,
            personas=personas,
        )

    def _parse_project_file(self, filepath: Path) -> Tuple[str, str, List[Milestone]]:
        """Parse the main project.md file."""
        if not filepath.exists():
            return "", "", []

        content = filepath.read_text()
        lines = content.split("\n")

        name = ""
        description_lines = []
        milestones = []
        current_section = None
        in_milestone = False
        current_milestone = {}

        for line in lines:
            # Parse title
            if line.startswith("# "):
                name = line[2:].strip()
                continue

            # Track sections
            if line.startswith("## "):
                current_section = line[3:].strip().lower()
                in_milestone = False
                continue

            # Parse milestones
            if current_section == "milestones":
                milestone_match = re.match(r"^### (M\d+):\s*(.+)$", line)
                if milestone_match:
                    if current_milestone:
                        milestones.append(self._create_milestone(current_milestone))
                    current_milestone = {
                        "id": milestone_match.group(1),
                        "name": milestone_match.group(2),
                        "description": "",
                        "epic_ids": [],
                    }
                    in_milestone = True
                    continue

                if in_milestone and line.strip():
                    # Parse epic references
                    epic_match = re.match(r"^-\s*\[([ x])\]\s*(E\d+)", line)
                    if epic_match:
                        current_milestone["epic_ids"].append(epic_match.group(2))
                    elif not line.startswith("-"):
                        current_milestone["description"] += line.strip() + " "
                continue

            # Parse description
            if current_section == "description" or (not current_section and line.strip()):
                description_lines.append(line)

        # Add final milestone
        if current_milestone:
            milestones.append(self._create_milestone(current_milestone))

        return name, "\n".join(description_lines).strip(), milestones

    def _create_milestone(self, data: dict) -> Milestone:
        """Create a Milestone from parsed data."""
        return Milestone(
            id=data["id"],
            name=data["name"],
            description=data.get("description", "").strip(),
            epic_ids=data.get("epic_ids", []),
        )

    def _parse_personas_dir(self, personas_dir: Path) -> List[Persona]:
        """Parse all persona files in a directory."""
        personas = []
        if not personas_dir.exists():
            return personas

        for filepath in personas_dir.glob("*.md"):
            persona = self._parse_persona_file(filepath)
            if persona:
                personas.append(persona)

        return personas

    def _parse_persona_file(self, filepath: Path) -> Optional[Persona]:
        """Parse a single persona markdown file."""
        content = filepath.read_text()
        lines = content.split("\n")

        persona_id = filepath.stem
        name = ""
        role = ""
        capabilities = []
        constraints = []
        claude_md_lines = []
        current_section = None
        in_template = False

        for line in lines:
            # Parse title
            if line.startswith("# "):
                name = line[2:].strip()
                continue

            # Track sections
            if line.startswith("## "):
                section_name = line[3:].strip().lower()
                if "template" in section_name or "claude" in section_name:
                    current_section = "template"
                    in_template = True
                else:
                    current_section = section_name
                    in_template = False
                continue

            # Parse role
            if current_section == "role" and line.strip():
                role = line.strip()
                continue

            # Parse capabilities
            if current_section == "capabilities":
                cap_match = re.match(r"^-\s*(.+)$", line)
                if cap_match:
                    capabilities.append(cap_match.group(1).strip())
                continue

            # Parse constraints
            if current_section == "constraints":
                const_match = re.match(r"^-\s*(.+)$", line)
                if const_match:
                    constraints.append(const_match.group(1).strip())
                continue

            # Collect template content
            if in_template:
                claude_md_lines.append(line)

        return Persona(
            id=persona_id,
            name=name,
            role=role,
            capabilities=capabilities,
            constraints=constraints,
            claude_md_template="\n".join(claude_md_lines).strip(),
        )

    def _parse_epics_and_stories(self, epics_dir: Path, stories_dir: Path) -> List[Epic]:
        """Parse all epics and their associated stories."""
        epics = []
        if not epics_dir.exists():
            return epics

        for filepath in epics_dir.glob("*.md"):
            epic = self._parse_epic_file(filepath)
            if epic:
                # Load stories for this epic
                epic_stories_dir = stories_dir / epic.id
                if epic_stories_dir.exists():
                    for story_file in epic_stories_dir.glob("*.md"):
                        story = self._parse_story_file(story_file, epic.id)
                        if story:
                            epic.stories.append(story)
                epics.append(epic)

        return epics

    def _parse_epic_file(self, filepath: Path) -> Optional[Epic]:
        """Parse a single epic markdown file."""
        content = filepath.read_text()
        lines = content.split("\n")

        # Extract ID from filename (E001-feature.md -> E001)
        epic_id_match = re.match(r"(E\d+)", filepath.stem)
        epic_id = epic_id_match.group(1) if epic_id_match else filepath.stem

        title = ""
        description_lines = []
        status = EpicStatus.NOT_STARTED
        milestone_id = None
        priority = "medium"
        dependencies = []
        current_section = None

        for line in lines:
            # Parse title
            if line.startswith("# "):
                title_match = re.match(r"^#\s*(?:E\d+:\s*)?(.+)$", line)
                if title_match:
                    title = title_match.group(1).strip()
                continue

            # Track sections
            if line.startswith("## "):
                current_section = line[3:].strip().lower()
                continue

            # Parse metadata
            if current_section == "metadata":
                status_match = re.match(r"^-\s*\*\*Status\*\*:\s*(\w+)", line, re.IGNORECASE)
                if status_match:
                    try:
                        status = EpicStatus(status_match.group(1).lower())
                    except ValueError:
                        pass
                    continue

                priority_match = re.match(r"^-\s*\*\*Priority\*\*:\s*(\w+)", line, re.IGNORECASE)
                if priority_match:
                    priority = priority_match.group(1).lower()
                    continue

                milestone_match = re.match(r"^-\s*\*\*Milestone\*\*:\s*(M\d+)", line, re.IGNORECASE)
                if milestone_match:
                    milestone_id = milestone_match.group(1)
                    continue

                deps_match = re.match(r"^-\s*\*\*Dependencies\*\*:\s*\[([^\]]*)\]", line, re.IGNORECASE)
                if deps_match:
                    deps_str = deps_match.group(1)
                    if deps_str.strip():
                        dependencies = [d.strip() for d in deps_str.split(",")]
                    continue

            # Parse description
            if current_section == "description" and line.strip():
                description_lines.append(line)

        return Epic(
            id=epic_id,
            title=title,
            description="\n".join(description_lines).strip(),
            status=status,
            milestone_id=milestone_id,
            priority=priority,
            dependencies=dependencies,
        )

    def _parse_story_file(self, filepath: Path, epic_id: str) -> Optional[Story]:
        """Parse a single story markdown file."""
        content = filepath.read_text()
        lines = content.split("\n")

        # Extract ID from filename (S001-login-flow.md -> S001)
        story_id_match = re.match(r"(S\d+)", filepath.stem)
        story_id = story_id_match.group(1) if story_id_match else filepath.stem

        title = ""
        status = StoryStatus.NOT_STARTED
        as_a = ""
        i_want = ""
        so_that = ""
        acceptance_criteria = []
        test_scenarios = []
        tasks = []
        dependencies = []
        blocks = []
        technical_notes_lines = []
        current_section = None
        ac_counter = 0
        ts_counter = 0
        task_counter = 0

        for line in lines:
            # Parse title
            if line.startswith("# "):
                title_match = re.match(r"^#\s*(?:S\d+:\s*)?(.+)$", line)
                if title_match:
                    title = title_match.group(1).strip()
                continue

            # Track sections
            if line.startswith("## "):
                current_section = line[3:].strip().lower()
                continue

            # Parse metadata
            if current_section == "metadata":
                status_match = re.match(r"^-\s*\*\*Status\*\*:\s*(\w+)", line, re.IGNORECASE)
                if status_match:
                    try:
                        status = StoryStatus(status_match.group(1).lower().replace(" ", "_"))
                    except ValueError:
                        pass
                    continue

                deps_match = re.match(r"^-\s*\*\*Dependencies\*\*:\s*\[([^\]]*)\]", line, re.IGNORECASE)
                if deps_match:
                    deps_str = deps_match.group(1)
                    if deps_str.strip():
                        dependencies = [d.strip() for d in deps_str.split(",")]
                    continue

                blocks_match = re.match(r"^-\s*\*\*Blocks\*\*:\s*\[([^\]]*)\]", line, re.IGNORECASE)
                if blocks_match:
                    blocks_str = blocks_match.group(1)
                    if blocks_str.strip():
                        blocks = [b.strip() for b in blocks_str.split(",")]
                    continue

            # Parse user story
            if current_section == "user story" and line.strip():
                # Match: As a **user**, I want **to do X** so that **benefit**
                story_match = re.match(
                    r"^As\s+a?\s*\*\*(.+?)\*\*,?\s*I\s+want\s+(?:to\s+)?\*\*(.+?)\*\*\s*(?:so\s+that\s+\*\*(.+?)\*\*)?",
                    line, re.IGNORECASE
                )
                if story_match:
                    as_a = story_match.group(1).strip()
                    i_want = story_match.group(2).strip()
                    so_that = story_match.group(3).strip() if story_match.group(3) else ""
                continue

            # Parse acceptance criteria
            if "acceptance" in current_section.lower():
                ac_match = re.match(r"^(\d+)\.\s+(.+)$", line)
                if ac_match:
                    ac_counter += 1
                    ac = self._parse_acceptance_criterion(
                        f"AC{ac_counter:03d}",
                        ac_match.group(2).strip()
                    )
                    acceptance_criteria.append(ac)
                continue

            # Parse test scenarios
            if "test" in current_section.lower() and "scenario" in current_section.lower():
                ts_match = re.match(r"^-\s+(\w+)\(\)(.*)$", line)
                if ts_match:
                    ts_counter += 1
                    ts = TestScenario(
                        id=f"TS{ts_counter:03d}",
                        name=ts_match.group(1),
                        description=ts_match.group(2).strip() if ts_match.group(2) else "",
                        acceptance_criterion_id=f"AC{ts_counter:03d}" if ts_counter <= ac_counter else "",
                        test_function_name=ts_match.group(1),
                    )
                    test_scenarios.append(ts)
                continue

            # Parse tasks
            if current_section == "tasks":
                task_match = re.match(r"^-\s*\[([ x])\]\s*(T\d+):\s*(.+)$", line)
                if task_match:
                    is_complete = task_match.group(1).lower() == "x"
                    task_id = task_match.group(2)
                    task_desc_and_meta = task_match.group(3)

                    # Parse task description and metadata
                    task = self._parse_task_line(task_id, task_desc_and_meta, is_complete)
                    tasks.append(task)
                continue

            # Parse technical notes
            if "technical" in current_section.lower():
                technical_notes_lines.append(line)

        return Story(
            id=story_id,
            title=title,
            epic_id=epic_id,
            status=status,
            as_a=as_a,
            i_want=i_want,
            so_that=so_that,
            acceptance_criteria=acceptance_criteria,
            test_scenarios=test_scenarios,
            tasks=tasks,
            dependencies=dependencies,
            blocks=blocks,
            technical_notes="\n".join(technical_notes_lines).strip(),
        )

    def _parse_acceptance_criterion(self, ac_id: str, text: str) -> AcceptanceCriterion:
        """Parse an acceptance criterion, looking for Given/When/Then."""
        given = ""
        when = ""
        then = ""

        # Try to parse Given/When/Then format
        gwt_match = re.match(
            r"Given\s+(.+?),?\s+(?:when\s+(.+?),?\s+)?(?:then\s+)?(.+)$",
            text, re.IGNORECASE
        )
        if gwt_match:
            given = gwt_match.group(1).strip()
            when = gwt_match.group(2).strip() if gwt_match.group(2) else ""
            then = gwt_match.group(3).strip()

        return AcceptanceCriterion(
            id=ac_id,
            description=text,
            given=given,
            when=when,
            then=then,
        )

    def _parse_task_line(self, task_id: str, line: str, is_complete: bool) -> Task:
        """Parse a task line with description and metadata."""
        # Format: Description | role:implementer | depends:T001,T002
        parts = [p.strip() for p in line.split("|")]
        description = parts[0]
        role = ""
        dependencies = []

        for part in parts[1:]:
            role_match = re.match(r"role:(\w+)", part, re.IGNORECASE)
            if role_match:
                role = role_match.group(1)
                continue

            deps_match = re.match(r"depends?:([\w,]+)", part, re.IGNORECASE)
            if deps_match:
                dependencies = [d.strip() for d in deps_match.group(1).split(",")]
                continue

        status = TaskStatus.COMPLETE if is_complete else TaskStatus.AVAILABLE

        return Task(
            id=task_id,
            description=description,
            role=role,
            status=status,
            dependencies=dependencies,
            created_at=datetime.now(),
        )

File: src/plan/parser.js
/**
 * @file Markdown parser for project plans.
 * Parses plan directories containing project.md and persona files.
 * @module plan/parser
 */

import { readFile, readdir, access, constants } from 'node:fs/promises';
import { join, basename } from 'node:path';
import { PlanParseError } from '../orchestrator/errors.js';
import {
  ProjectPlan,
  Epic,
  Story,
  Task,
  Milestone,
  Persona,
  AcceptanceCriterion,
} from './models.js';
import { TaskStatus, StoryStatus, EpicStatus } from '../types/index.js';

/**
 * Parser for project plan markdown files.
 */
export class PlanParser {
  /**
   * Create a PlanParser.
   * @param {string} planDir - Directory containing plan files
   */
  constructor(planDir) {
    /** @type {string} */
    this.planDir = planDir;
  }

  /**
   * Parse the complete plan from the directory.
   * @returns {Promise<ProjectPlan>}
   * @throws {PlanParseError} If parsing fails
   */
  async parsePlan() {
    try {
      // Check if plan directory exists
      await access(this.planDir, constants.R_OK);
    } catch {
      throw new PlanParseError(`Plan directory not found: ${this.planDir}`);
    }

    // Parse main project file
    const projectPath = join(this.planDir, 'project.md');
    let projectContent;
    try {
      projectContent = await readFile(projectPath, 'utf-8');
    } catch {
      throw new PlanParseError(`Project file not found: ${projectPath}`, { file: projectPath });
    }

    const { name, description } = this._parseProjectFile(projectContent);

    // Parse personas
    const personas = await this._parsePersonasDir();

    // Parse epics and stories
    const epics = await this._parseEpicsDir();

    // Parse milestones
    const milestones = await this._parseMilestonesFile();

    return new ProjectPlan({
      name,
      description,
      epics,
      milestones,
      personas,
    });
  }

  /**
   * Parse the main project.md file.
   * @private
   * @param {string} content - File content
   * @returns {{name: string, description: string}}
   */
  _parseProjectFile(content) {
    const lines = content.split('\n');
    let name = '';
    let description = '';
    let inDescription = false;

    for (const line of lines) {
      if (line.startsWith('# ')) {
        name = line.slice(2).trim();
        inDescription = true;
      } else if (inDescription && line.startsWith('## ')) {
        break;
      } else if (inDescription) {
        description += line + '\n';
      }
    }

    return {
      name,
      description: description.trim(),
    };
  }

  /**
   * Parse personas from the personas directory.
   * @private
   * @returns {Promise<Persona[]>}
   */
  async _parsePersonasDir() {
    const personasDir = join(this.planDir, 'personas');
    const personas = [];

    try {
      await access(personasDir, constants.R_OK);
      const files = await readdir(personasDir);

      for (const file of files) {
        if (!file.endsWith('.md')) continue;

        const content = await readFile(join(personasDir, file), 'utf-8');
        const persona = this._parsePersonaFile(content, file);
        if (persona) {
          personas.push(persona);
        }
      }
    } catch {
      // Personas directory is optional
    }

    return personas;
  }

  /**
   * Parse a single persona file.
   * @private
   * @param {string} content - File content
   * @param {string} filename - File name
   * @returns {Persona|null}
   */
  _parsePersonaFile(content, filename) {
    const lines = content.split('\n');
    const id = basename(filename, '.md');
    let name = '';
    let role = '';
    const capabilities = [];
    const constraints = [];
    let claudeMdTemplate = '';

    let currentSection = '';

    for (const line of lines) {
      if (line.startsWith('# ')) {
        name = line.slice(2).trim();
      } else if (line.startsWith('## Role')) {
        currentSection = 'role';
      } else if (line.startsWith('## Capabilities')) {
        currentSection = 'capabilities';
      } else if (line.startsWith('## Constraints')) {
        currentSection = 'constraints';
      } else if (line.startsWith('## Template')) {
        currentSection = 'template';
      } else if (line.startsWith('## ')) {
        currentSection = '';
      } else if (currentSection === 'role' && line.trim()) {
        role = line.trim();
        currentSection = '';
      } else if (currentSection === 'capabilities' && line.startsWith('- ')) {
        capabilities.push(line.slice(2).trim());
      } else if (currentSection === 'constraints' && line.startsWith('- ')) {
        constraints.push(line.slice(2).trim());
      } else if (currentSection === 'template') {
        claudeMdTemplate += line + '\n';
      }
    }

    if (!name) return null;

    return new Persona({
      id,
      name,
      role: role || id,
      capabilities,
      constraints,
      claudeMdTemplate: claudeMdTemplate.trim(),
    });
  }

  /**
   * Parse epics from the epics directory.
   * @private
   * @returns {Promise<Epic[]>}
   */
  async _parseEpicsDir() {
    const epicsDir = join(this.planDir, 'epics');
    const epics = [];

    try {
      await access(epicsDir, constants.R_OK);
      const files = await readdir(epicsDir);

      for (const file of files) {
        if (!file.endsWith('.md')) continue;

        const content = await readFile(join(epicsDir, file), 'utf-8');
        const epic = this._parseEpicFile(content, file);
        if (epic) {
          epics.push(epic);
        }
      }
    } catch {
      // Epics directory is optional
    }

    return epics;
  }

  /**
   * Parse a single epic file.
   * @private
   * @param {string} content - File content
   * @param {string} filename - File name
   * @returns {Epic|null}
   */
  _parseEpicFile(content, filename) {
    const lines = content.split('\n');
    const id = basename(filename, '.md');
    let title = '';
    let description = '';
    let milestoneId = null;
    let priority = 'medium';
    const dependencies = [];
    const stories = [];

    let currentSection = '';
    let currentStory = null;
    let currentTask = null;

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];

      if (line.startsWith('# ')) {
        title = line.slice(2).trim();
        currentSection = 'description';
      } else if (line.startsWith('## Description')) {
        currentSection = 'description';
      } else if (line.startsWith('## Milestone:')) {
        milestoneId = line.split(':')[1]?.trim() || null;
      } else if (line.startsWith('## Priority:')) {
        priority = line.split(':')[1]?.trim() || 'medium';
      } else if (line.startsWith('## Dependencies')) {
        currentSection = 'dependencies';
      } else if (line.startsWith('## Story:') || line.startsWith('### Story:')) {
        // Save previous story
        if (currentStory) {
          if (currentTask) {
            currentStory.tasks.push(currentTask);
            currentTask = null;
          }
          stories.push(currentStory);
        }

        const storyTitle = line.split(':').slice(1).join(':').trim();
        const storyId = `${id}-S${stories.length + 1}`;
        currentStory = new Story({
          id: storyId,
          title: storyTitle,
          epicId: id,
        });
        currentSection = 'story';
      } else if (line.startsWith('### Task:') || line.startsWith('#### Task:')) {
        // Save previous task
        if (currentTask && currentStory) {
          currentStory.tasks.push(currentTask);
        }

        const taskParts = line.split(':').slice(1).join(':').trim();
        const taskMatch = taskParts.match(/\[([^\]]+)\]\s*(.*)/);
        const role = taskMatch ? taskMatch[1] : 'developer';
        const taskDesc = taskMatch ? taskMatch[2] : taskParts;
        const taskId = currentStory ? `${currentStory.id}-T${currentStory.tasks.length + 1}` : `T${i}`;

        currentTask = new Task({
          id: taskId,
          description: taskDesc,
          role,
        });
        currentSection = 'task';
      } else if (currentSection === 'description' && !line.startsWith('##')) {
        description += line + '\n';
      } else if (currentSection === 'dependencies' && line.startsWith('- ')) {
        dependencies.push(line.slice(2).trim());
      } else if (currentSection === 'story' && line.startsWith('As a ')) {
        // Parse user story format
        const storyMatch = line.match(/As a ([^,]+), I want ([^,]+), so that (.+)/i);
        if (storyMatch && currentStory) {
          currentStory.asA = storyMatch[1].trim();
          currentStory.iWant = storyMatch[2].trim();
          currentStory.soThat = storyMatch[3].trim();
        }
      } else if (currentSection === 'story' && line.startsWith('- AC:')) {
        // Parse acceptance criterion
        if (currentStory) {
          const acText = line.slice(5).trim();
          const acId = `${currentStory.id}-AC${currentStory.acceptanceCriteria.length + 1}`;
          currentStory.acceptanceCriteria.push(
            new AcceptanceCriterion({
              id: acId,
              description: acText,
            })
          );
        }
      }
    }

    // Save last story and task
    if (currentTask && currentStory) {
      currentStory.tasks.push(currentTask);
    }
    if (currentStory) {
      stories.push(currentStory);
    }

    if (!title) return null;

    return new Epic({
      id,
      title,
      description: description.trim(),
      milestoneId,
      priority,
      dependencies,
      stories,
    });
  }

  /**
   * Parse milestones from the milestones.md file.
   * @private
   * @returns {Promise<Milestone[]>}
   */
  async _parseMilestonesFile() {
    const milestonesPath = join(this.planDir, 'milestones.md');
    const milestones = [];

    try {
      const content = await readFile(milestonesPath, 'utf-8');
      const lines = content.split('\n');

      let currentMilestone = null;

      for (const line of lines) {
        if (line.startsWith('## ')) {
          if (currentMilestone) {
            milestones.push(currentMilestone);
          }

          const title = line.slice(3).trim();
          const idMatch = title.match(/\[([^\]]+)\]/);
          const id = idMatch ? idMatch[1] : `M${milestones.length + 1}`;
          const name = title.replace(/\[[^\]]+\]/, '').trim();

          currentMilestone = new Milestone({
            id,
            name,
            description: '',
          });
        } else if (currentMilestone) {
          if (line.startsWith('- Epic:')) {
            const epicId = line.split(':')[1]?.trim();
            if (epicId) {
              currentMilestone.epicIds.push(epicId);
            }
          } else if (line.startsWith('- Target:')) {
            const dateStr = line.split(':').slice(1).join(':').trim();
            try {
              currentMilestone.targetDate = new Date(dateStr);
            } catch {
              // Invalid date, ignore
            }
          } else if (line.trim() && !line.startsWith('-')) {
            currentMilestone.description += line + '\n';
          }
        }
      }

      if (currentMilestone) {
        milestones.push(currentMilestone);
      }
    } catch {
      // Milestones file is optional
    }

    return milestones;
  }

  /**
   * Parse a plan from JSON file.
   * Alternative to markdown parsing.
   * @param {string} jsonPath - Path to JSON file
   * @returns {Promise<ProjectPlan>}
   */
  async parseFromJson(jsonPath) {
    try {
      const content = await readFile(jsonPath, 'utf-8');
      const data = JSON.parse(content);
      return ProjectPlan.fromDict(data);
    } catch (error) {
      throw new PlanParseError(`Failed to parse JSON plan: ${error.message}`, {
        file: jsonPath,
        cause: error,
      });
    }
  }
}
